ex2_hal/iris/hardware_interface/source/iris.c

```diff
@@ -0,0 +1,460 @@
+            {
+                ret = send_command(IRIS_TRANSFER_IMAGE);
+                if (ret == IRIS_ACK) {
+                    controller_state = GET_DATA;
+                } else {
+                    controller_state = FINISH;
+                }
+                break;
+            }
+            case GET_DATA: // Get image data in chunks/blocks
+            {
+                uint16_t * image_data_buffer = (uint16_t*) pvPortMalloc(IMAGE_TRANSFER_SIZE * sizeof(uint16_t));
+                memset(image_data_buffer, 0, IMAGE_TRANSFER_SIZE);
+                if (image_data_buffer == NULL) {
+                    ex2_log("Failed attempt to dynamically allocate memory under iris get image data");
+                    return IRIS_HAL_ERROR;;
+                }
+                num_transfer = (IMAGE_TRANSFER_SIZE + image_length) / IMAGE_TRANSFER_SIZE; // Ceiling division
+                for (uint32_t count_transfer = 0; count_transfer < num_transfer; count_transfer++) {
+                    ret = get_data(image_data_buffer, IMAGE_TRANSFER_SIZE);
+                    // TODO: Do something with the received data (e.g transfer it to the SD card)
+                    // Or just get the data and send it forward to the next stage. Prefer not to have too
+                    // much data processing in driver code
+
+                    memset(image_data_buffer, 0, IMAGE_TRANSFER_SIZE);
+                }
+                free(image_data_buffer);
+                controller_state = FINISH;
+                break;
+            }
+            case FINISH:
+            {
+                ex2_log("Iris returns ACK on transfer image command");
+                return IRIS_HAL_OK;
+            }
+            case ERROR_STATE:
+            {
+                ex2_log("Iris returns NACK on transfer image command");
+                return IRIS_HAL_ERROR;
+            }
+        }
+    }
+    return IRIS_HAL_ERROR;
+}
+
+/**
+ * @brief
+ *   Sends a get image count command to Iris, and expects to receive the
+ *   number of images stored on iris
+ *
+ * @return
+ *   Returns the number of images stored on Iris
+ **/
+IrisHALReturn iris_get_image_count(uint8_t *image_count) {
+    IrisLowLevelReturn ret;
+
+    controller_state = SEND_COMMAND;
+
+    while(controller_state != FINISH && controller_state != ERROR_STATE) {
+        switch (controller_state) {
+            case SEND_COMMAND:
+            {
+                ret = send_command(IRIS_GET_IMAGE_COUNT);
+                if (ret == IRIS_ACK) {
+                    controller_state = GET_DATA;
+                } else {
+                    controller_state = FINISH;
+                }
+                break;
+            }
+            case GET_DATA:
+            {
+                //TODO: write code to receive image count
+                break;
+            }
+            case FINISH:
+            {
+                ex2_log("Iris returns ACK on transfer image command");
```

Comment:
Use sys_log

Reply:
What is the difference between sys_log and ex2_log

Reply:
Should be implemented, although I will need to think a bit more on where to use the log in hardware interface or service. This will be done in a new PR as this is part of the error handling user story